proposals/lambdas-05-traits.md

```diff
@@ -0,0 +1,72 @@
+Now let's try to translate those "Kinds" into traits that fit into the type system:
+
+```savi
+// `->` (Kind A)
+// (cannot be objectified, so does not need to have a corresponding trait)
+
+// `...->` FunOnce'iso (Kind B)
+:trait iso FunOnce(R, P1, P2)
+  :fun iso call_once(P1, P2) R
+
+// `...->` Fun'iso (Kind C)
+// `...->` Fun'ref (Kind D)
+:trait ref Fun(R, P1, P2)
+  :fun ref call(p1 P1, p2 P2) R
+
+  // Fun'iso can be a subtype of FunOnce'iso thanks to this wrapper method:
+  :fun iso call_once(p1 P1, p2 P2): @call(p1, p2)
+
+// `...->` FunVal'val (Kind E)
+:trait val FunVal(R, P1, P2)
+  :fun val call(P1, P2) R
+
+// `...->` FunVal'val (Kind F)
+// (note that in Savi, `non` types can adhere to a `val` trait/fun,
+// so it's okay to put stateless functions on `non` types here as `FunVal'val` too)
+```
+
+Notice that these traits (`FunVal`, `Fun`, `FunOnce`, respectively) mostly mirror the hierarchy of function traits in Rust (`Fn <: FnMut <: FnOnce`), except that in our type system `FunVal` cannot generally be a subtype of `Fun` because the former is `:fun val` and the latter is `:fun ref`. However `:fun box` can satisfy a `:fun ref` trait, so meeting the `Fun` trait does not strictly require a mutating function. Hence, the mutation-compatible one gets naming primacy in Savi (Savi `Fun` ~= Rust `FnMut`, and Savi `FunVal` =~ Rust `Fn`)
+
+We mostly make up for deficiency of `FunVal` not being a subtype of `Fun` by relying on type anonymity to allow a `FunVal`-compatible `...->` block literal to be objectified as one of the other trait types as long as it happens immediately at the creation site. That is, we can conceptually treat the immutable block as being a sub-object inside a mutable one, as long as we know this at the definition site so that we can define the anonymous type differently; and because the type is anonymous, it is not instantiable in any other place, so no other code can notice the discrepancy in how it was defined.
```

Comment:
So far I haven't defined any way of creating a raw lambda without it being passed to some function. I actually view this limitation to be a positive feature, because it forces specific typing of all lambda objects and we shouldn't have to deal with ambiguity on how a particular block should be treated.

That is, the function you pass it to will define the type of lambda it expects to receive, and then the `...->` block you pass to it will have to meet those requirements, and will be objectified with the relevant trait and ref cap.

Note that there is also not a way in the syntax to specify the parameter types of a `...->` block at the call site. This as well is expected to be inferred from looking at the signature of the function that receives and objectifies it.

I think this is a good tradeoff because it completely removes the ambiguity in typing while keeping the type information away from being a burden on the call site.

In cases where you don't actually have a need to pass it to a function immediately, you can have named stub functions that only exist to receive a `...->` block with a specific type and immediately return it. These stub functions would be named in such a way as to help the reader understand how this lambda was to be used later, such as in the following example, where `FileWatcher.change_callback` is such a stub function (a `:fun non`, callable like a static function) that only exists for the purposes of assigning a type to the block:

```savi
on_change = FileWatcher.change_callback ...-> (change_info |
  // (do something with `change_info`)
)

// ... some time later, or in another function

my_file_watcher.on_change = on_change
```

I actually think this is quite nice way of avoiding ambiguity while preserving and even enhancing readability.